### PR TITLE
[Merged by Bors] - Adjust beacon node timeouts for validator client HTTP requests

### LINKED ISSUE #2313: HTTP timeout too long for VC attestation duty
## Description

I recently noticed a `CRIT` error on a VC (sorry, I lost the log) that expressed the following:

1. The primary BN timed out when trying to produce an attestation with `/eth/v1/validator/attestation_data`
1. The secondary BN failed with [`AttestingPriorToHead`](https://github.com/sigp/lighthouse/blob/3a24ca5f14c6e6d6612fd43eca82aa0c1e6aba16/beacon_node/beacon_chain/src/beacon_chain.rs#L946-L949)

I believe this is because the timeout for HTTP calls on the VC is too long at 12s (one slot):

https://github.com/sigp/lighthouse/blob/3a24ca5f14c6e6d6612fd43eca82aa0c1e6aba16/validator_client/src/lib.rs#L59

Considering that attestation production happens 4 seconds into the slot, we should really have a timeout that's *less than* `12 - 4 == 8` seconds. I think an even smaller timeout would be better for attestations, they're very fast to produce and I think we should really be aborting if it takes longer that 1-2s.

However, such a short time doesn't work well for block production which is rather heavy. As such, I think the best way to solve this would be to allow for different timeouts for different requests.

## Version

`v1.3.0`: https://github.com/sigp/lighthouse/commit/3a24ca5f14c6e6d6612fd43eca82aa0c1e6aba16

### PR DESCRIPTION
**Pull request by macladson**
## Issue Addressed

Resolves #2313 

## Proposed Changes

Provide `BeaconNodeHttpClient` with a dedicated `Timeouts` struct.
This will allow granular adjustment of the timeout duration for different calls made from the VC to the BN. These can either be a constant value, or as a ratio of the slot duration.

Improve timeout performance by using these adjusted timeout duration's only whenever a fallback endpoint is available.

Add a CLI flag called `use-long-timeouts` to revert to the old behavior.

## Additional Info

Additionally set the default `BeaconNodeHttpClient` timeouts to the be the slot duration of the network, rather than a constant 12 seconds. This will allow it to adjust to different network specifications.

### CODE REVIEW

**Comment by macladson** [comment]
A few additional notes:
- As per the `reqwest` documentation, timeout duration can be adjusted on a [per-request basis](https://docs.rs/reqwest/0.11.0/reqwest/struct.RequestBuilder.html#method.timeout) without overwriting the default client timeout.

- We should probably have a discussion around what kind of timeouts are desired here to ensure proper fallback behavior. ~As a start I've put in a ratio of 3 (which is a 4 second timeout on mainnet spec networks) for all calls relevant to attestation production.~

- Something we should also keep in mind is the use of remote beacon nodes (like Infura) and needing to account for
the additional latency involved there. A timeout suitable for a local node, might not be suitable for a remote one.
To see the sorts of constraints we are working with, I ran this overnight using an Infura beacon node with a 1 second timeout per attestation request and didn't have any attestation timeouts, so I suspect we will generally be fine.

**Comment by macladson** [comment]
Just had an attestation timeout in the `eth1-simulator-ubuntu` CI test the first time it ran (but not the second). Not sure if it was a once off issue or if it means that this specific test isn't consistently able to handle shorter timeouts.

File: common/eth2/src/lib.rs
Comment (paulhauner) [inline common/eth2/src/lib.rs]: I would be tempted to do away with this `new` function and make the fields `pub` so it can be initialized "manually".

Whenever there's a string of identical types going into a function I worry about making a mistake with the ordering. I feel that ...

```rust
Thing {
  x: 1,
  y: 2,
  z: 3
}
```

... reads a bit better than ...

```rust
Thing::new(1, 2, 3)
```

... especially when the `new` function isn't doing anything else.

It's stylistic suggestion, feel free to do as you see fit :)

File: common/eth2/src/lib.rs
Comment (paulhauner) [inline common/eth2/src/lib.rs]: Another stylistic suggestion; I'd avoid using `default`. I feel the [`Default`](https://doc.rust-lang.org/std/default/trait.Default.html) trait already has a monopoly on that function name.

Suggestion: `pub fn all(duration: Duration)`. Not in love with that name though..

Once again, just a suggestion.

File: validator_client/src/cli.rs
Comment (paulhauner) [inline validator_client/src/cli.rs]: To help nudge users in the right direction, perhaps we could append:

> This flag is generally not recommended, longer timeouts can caused missed duties when fallbacks are used.

File: validator_client/src/lib.rs
Comment (paulhauner) [inline validator_client/src/lib.rs]: The [`Duration::from_secs`](https://doc.rust-lang.org/std/time/struct.Duration.html#method.from_secs) and other functions are `const`, so you can do this:

```rust
const HTTP_ATTESTATION_TIMEOUT: Duration = Duration::from_secs(1);
```

Same goes for `HTTP_ATTESTER_DUTIES_TIMEOUT` below.

File: validator_client/src/lib.rs
Comment (paulhauner) [inline validator_client/src/lib.rs]: I see what you're doing here and I think it works, but I do notice that we're satisfying *"is this the last item in the list"* by asking *"is this the same value as the last item in the list"*. Perhaps we're better served by a solution that doesn't make assumptions about the list containing duplicates.

Here's a suggestion:

```rust
let last_index = config
  .beacon_nodes
  .len()
  .checked_sub(1)
  .ok_or_else(|| "No beacon nodes defined.".to_string())?;

config
  .beacon_nodes
  .iter()
  .enumerate()
  .map(|(i, url)| {
      if i < last_index {
        // smol timeouts
      } else {
        // chonky timeouts
      }
  })
```

**Review by paulhauner — CHANGES_REQUESTED** [review]
Heyo! Looking good, I'm sure this would work fine as-is. I must made a few nitpicks.

I've also added some metrics via this commit: https://github.com/sigp/lighthouse/commit/0e7ac92e1179d16333e30c51ee9a0dd43b9e601c I'd suggest cherry-picking it here so we can have them in the future too.

I'm running some metrics on a couple of Prater nodes. Here's the info so far:

http_times

We can let it run over the weekend and see what sort of times we're seeing on those requests. From an initial glance it looks like your suggestions are pretty good. I'll give you access to the metrics nodes when were both online again (probably Monday).

File: validator_client/src/lib.rs
Comment (macladson) [reply validator_client/src/lib.rs]: Ah yes, good call. This is definitely a smarter way to do this.

**Comment by macladson** [comment]
Thanks for the suggestions, and for the additional metrics! Let me know if you feel the timeout durations need any more adjustment.

**Comment by paulhauner** [comment]
Here's the latest times, over 2 days on Prater:

http_times2


I'll make a commit with some suggestions based off this and drop it here :slightly_smiling_face:

**Comment by paulhauner** [comment]
Actually, I think this view is better. It's the maximum times during the last 2 days on Prater:

http_times_vals

**Comment by paulhauner** [comment]
Here's my suggested timeouts :slightly_smiling_face: Please double check and let me know if I've done anything stupid :pray: 

https://github.com/paulhauner/lighthouse/commit/02e4b82662dd66857cb3f9f2ef9365062e0b64fe

**Comment by macladson** [comment]
Looks good to me! I'll pull it in

**Comment by paulhauner** [comment]
I'm running this on our European Prater nodes. After that's been running a while, I'll merge this in :)

**Comment by paulhauner** [comment]
This ran well on the testnet nodes! I'd like to check with @michaelsproul before merging. We *might* want to block this on #2279.

**Comment by paulhauner** [comment]
#2279 should merge today or early next week, we can merge this after :slightly_smiling_face:

**Comment by paulhauner** [comment]
#2279 has merged! :tada: No conflicts visible yet, let's try bors.

bors r+

**Review by paulhauner — APPROVED** [review]

**Pull request closed without merging**
